compiler/qsc_frontend/src/typeck/tests/bounded_polymorphism.rs

```diff
@@ -0,0 +1,822 @@
+        "##]],
+    );
+}
+
+#[test]
+fn exp() {
+    check(
+        r#"
+        namespace A {
+            function Foo<'T: Exp[Int]>(a: 'T, b: Int) : 'T {
+                a ^ b
+            }
+        }
+        "#,
+        "",
+        &expect![[r##"
+            #11 61-76 "(a: 'T, b: Int)" : (Param<"'T": 0>, Int)
+            #12 62-67 "a: 'T" : Param<"'T": 0>
+            #16 69-75 "b: Int" : Int
+            #23 82-119 "{\n                a ^ b\n            }" : Param<"'T": 0>
+            #25 100-105 "a ^ b" : Param<"'T": 0>
+            #26 100-101 "a" : Param<"'T": 0>
+            #29 104-105 "b" : Int
+        "##]],
+    );
+}
+
+#[test]
+fn exp_fail() {
+    check(
+        r#"
+        namespace A {
+            function Foo<'T: Exp[Int]>(a: 'T, b: Bool) : 'T {
+                a ^ b
+            }
+        }
+        "#,
+        "",
+        &expect![[r##"
+            #11 61-77 "(a: 'T, b: Bool)" : (Param<"'T": 0>, Bool)
+            #12 62-67 "a: 'T" : Param<"'T": 0>
+            #16 69-76 "b: Bool" : Bool
+            #23 83-120 "{\n                a ^ b\n            }" : Param<"'T": 0>
+            #25 101-106 "a ^ b" : Param<"'T": 0>
+            #26 101-102 "a" : Param<"'T": 0>
+            #29 105-106 "b" : Bool
+            Error(Type(Error(TyMismatch("Int", "Bool", Span { lo: 101, hi: 106 }))))
+        "##]],
+    );
+}
+#[test]
+fn extra_arg_to_exp() {
+    check(
+        r#"
+        namespace A {
+            function Foo<'E, 'T: Exp['E, Int]>(a: 'E, b: Int) : 'T {
+                a ^ b
+            }
+            function Foo2<'E, 'T: Exp>(a: 'E, b: Int) : 'T {
+                a ^ b
+            }
+        }
+        "#,
+        "",
+        &expect![[r##"
+            #14 69-84 "(a: 'E, b: Int)" : (Param<"'E": 0>, Int)
+            #15 70-75 "a: 'E" : Param<"'E": 0>
+            #19 77-83 "b: Int" : Int
+            #26 90-127 "{\n                a ^ b\n            }" : Param<"'E": 0>
+            #28 108-113 "a ^ b" : Param<"'E": 0>
+            #29 108-109 "a" : Param<"'E": 0>
+            #32 112-113 "b" : Int
+            #41 166-181 "(a: 'E, b: Int)" : (Param<"'E": 0>, Int)
+            #42 167-172 "a: 'E" : Param<"'E": 0>
+            #46 174-180 "b: Int" : Int
+            #53 187-224 "{\n                a ^ b\n            }" : Param<"'E": 0>
+            #55 205-210 "a ^ b" : Param<"'E": 0>
+            #56 205-206 "a" : Param<"'E": 0>
+            #59 209-210 "b" : Int
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 2, span: Span { lo: 56, hi: 59 } })))
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 2, span: Span { lo: 56, hi: 59 } })))
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 0, span: Span { lo: 162, hi: 165 } })))
+            Error(Type(Error(IncorrectNumberOfConstraintParameters { expected: 1, found: 0, span: Span { lo: 162, hi: 165 } })))
```

Comment:
Wasn't this the duplicate error bug you said you fixed? Or was there more than one and one of them was benign?

Reply:
I think this one is going to be a fast-follow, I'd rather get the feature in for the release and track this one down shortly after. That being said, I'll track this down as soon as I'm done with the other things on this branch...so maybe I'll finish before it goes in. Tl;dr I'd rather this not block the PR, and I'll fix it ASAP, whether that be before or after it goes in.

Reply:
Sounds good!